src/client/tile.cpp

```diff
@@ -775,205 +770,79 @@ void TextureSource::rebuildImagesAndTextures()
-
-		video::ITexture *rtt = driver->addTexture(params.rtt_texture_name.c_str(), inventory_image);
-		inventory_image->drop();
-
-		if (rtt == NULL) {
-			errorstream << "TextureSource::generateTextureFromMesh(): failed to recreate texture from image: " << params.rtt_texture_name << std::endl;
-			return NULL;
-		}
-
-		driver->makeColorKeyTexture(rtt, v2s32(0,0));
-
-		if (params.delete_texture_on_shutdown)
-			m_texture_trash.push_back(rtt);
-
-		return rtt;
-	}
-#endif
-
-	if (!driver->queryFeature(video::EVDF_RENDER_TO_TARGET)) {
-		static bool warned = false;
-		if (!warned)
-		{
-			errorstream<<"TextureSource::generateTextureFromMesh(): "
-				<<"EVDF_RENDER_TO_TARGET not supported."<<std::endl;
-			warned = true;
+	video::IVideoDriver *driver = RenderingEngine::get_video_driver();
+	video::IImage *result = driver->createImage(video::ECF_A8R8G8B8, {size, size});
+	result->fill(video::SColor(0x00000000u));
+
+	// Offset of the view space origin (from the top image edge),
+	// where the upper near cube vertex is. Full image is 2 units.
+	static constexpr float base_y = 0.9f;
+
+	// Cube dimensions. The image is 2x2 in this (view) space.
+	static constexpr float diag_x = 0.8f;
+	static constexpr float diag_y = 0.4f;
+	static constexpr float vert_y = 1.2f;
+
+	static constexpr float coef_x = 0.5f / diag_x;
+	static constexpr float coef_y = -0.5f / diag_y;
+	static constexpr float coef_z = -1 / vert_y;
+
+	float scale = 2.f / size;
+	video::SColor *pixel = reinterpret_cast<video::SColor *>(result->lock());
+	for (int j = 0; j < size; j++) {
+		for (int i = 0; i < size; i++) {
+			// View space coordinates
+			float x = scale * (i + 0.5f) - 1.f;
+			float y = scale * (j + 0.5f) - base_y;
+
+			float u = coef_y * y + coef_x * x;
+			float v = coef_y * y - coef_x * x;
+
+			// Texture coordinates for cube faces, at current pixel.
+			v2f tex_top = {1.f - v, 1.f - u};
+			v2f tex_left = {2 * coef_x * x + 1.f, coef_z * u};
+			v2f tex_right = {2 * coef_x * x, coef_z * v};
+
+			// Draw from `src`, if current pixel is inside the corresponding quad,
+			// i.e. if texture coordinates are in range [0, 1).
+			// `b` stands for "brightness."
+			auto try_paint = [pixel] (video::IImage *src, core::dimension2du size, v2f uv, float b) -> void {
+				int k = size.Width * uv.X;
+				int l = size.Height * uv.Y;
+				if (k < 0 || k >= size.Width || l < 0 || l >= size.Height)
+					return;
+				video::SColor s = src->getPixel(k, l);
+				pixel->set(s.getAlpha(), b * s.getRed(), b * s.getGreen(), b * s.getBlue());
+			};
+
+			// Brightness is the same as in applyFacesShading
+			try_paint(top, size_top, tex_top, 1.000000f);
+			try_paint(left, size_left, tex_left, 0.836660f);
+			try_paint(right, size_right, tex_right, 0.670820f);
+
+			// Reference next pixel in the row or, if the row (and thus the inner loop)
+			// is over, first pixel of the next row (assuming there is no padding, but
+			// there shouldn't be any as row size is exact power of 2).
```

Comment:
looking at `IImage.h` it seems Irrlicht does not support having row padding anyway

Reply:
Not sure where in `IImage.h` you could find that; `IImage` has the `getPitch` method that allows handling images with padding.
OTOH, `CImage` doesn’t support any padding, using image width in internal calculations.

Instead of assuming anything, it may be better to add a `sanity_check` of padding absence.